README.md
# rosmon

![screenshot](https://github.com/xqms/rosmon/raw/master/doc/screenshot.png)

`rosmon` is a drop-in replacement for the standard `roslaunch` tool. Rather
unlike `roslaunch`, `rosmon` is focused on (remote) process monitoring.

Please see the [ROS wiki page](https://wiki.ros.org/rosmon) for further
information, the rest of this README contains information for `rosmon`
developers.

## Building

Simple include this repository in your catkin workspace. After a build
(tested with `catkin_tools`) and re-sourcing of the `devel/setup.bash` you will
have the `mon` command in your environment.

## License

`rosmon` is licensed under BSD-3. The `rosmon` repository includes the
[fmt](http://fmtlib.net/latest/index.html) library, which is also licensed
under BSD-2.

## Author

Max Schwarz <max.schwarz@ais.uni-bonn.de>


CHANGELOG.rst
^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Changelog for package rosmon
^^^^^^^^^^^^^^^^^^^^^^^^^^^^

1.0.10 (2018-10-29)
-------------------
* main: exit gracefully on SIGTERM and SIGHUP as well (issue #59, PR #60)
* launch: allow <arg default="XY"> in <include> tags (issue #57, PR #58)
  roslaunch allows this, so we should as well. Since it seems a bit
  confusing, we issue a warning when this happens.
  In these cases, <arg value="XY"> should be applicable and is much clearer.
* Contributors: Max Schwarz

1.0.9 (2018-09-20)
------------------
* Fix race condition in integration test (issue #42, PR #54)
* Clean namespace names properly to fix double slashes (issue #49, PR #53)
* Respect ROS_NAMESPACE for nested launches (issue #46, PR #51)
* gui: Sort Memory column correctly (issue #48, PR #50)
* Try to find an executable *file* for nodes (issue #45, PR #47)
* Add --stop-timeout option and launch file attribute (PR #37)
* Handle params with leading slashes inside nodes (PR #40)
* Add --no-start option (PR #39)
* gui: Fix index bug in showContextMenu (PR #38)
* Contributors: Max Schwarz, Nikos Skalkotos, Romain Reignier

1.0.8 (2018-08-07)
------------------
* main: Add option for flushing the logfile
  Add --flush option that will flush the logfile after each entry.
* Merge pull request `#35 <https://github.com/xqms/rosmon/issues/35>`_
  More complete support for rosparam features
* Merge pull request `#34 <https://github.com/xqms/rosmon/issues/34>`_
  YAML quoted strings
* launch: rosparam: support binary data
* launch: rosparam: correctly handle explicit YAML type tags
* launch: support rosparam angle computations
* launch: always map YAML quoted values to string params
  These always get mapped to str by python's yaml.load, which is used by
  roslaunch, so we do the same here.
* Contributors: Max Schwarz, Nikos Skalkotos

1.0.7 (2018-05-27)
------------------
* Support Python 3 & select appropriate Python version.
  This fixes a mismatch on Debian Jessie, where rospack is linked against
  Python 2.7 and we would link against Python 3.4.
* launch: substitution_python: support Python 3
* Contributors: Max Schwarz

1.0.6 (2018-05-26)
------------------
* test/xml: replace more multiline string Catch captures
  Sorry, somehow these slipped through - and I didn't have a good way of
  testing these locally without the buildfarm. I'm testing with gcc 4.8
  on trusty now, which seems to have matching behavior.
* Contributors: Max Schwarz

1.0.5 (2018-05-25)
------------------
* test/xml: avoid multiline Catch captures in exception tests
  These trigger some weird bug between g++ 5.4 and ccache, which is used
  in the ROS buildfarm. [...]
  This should fix compilation on the build farm.
* Contributors: Max Schwarz

1.0.4 (2018-05-24)
------------------
* Merge pull request `#30 <https://github.com/xqms/rosmon/issues/30>`_ from xqms/feature/global_remap
  Support <remap> in other scopes than <node>. Fixes `#28 <https://github.com/xqms/rosmon/issues/28>`_.
* Merge pull request `#29 <https://github.com/xqms/rosmon/issues/29>`_ from xqms/feature/fmt
  Port all string formatting to fmt
* update README.md, refer to ROS wiki
  Otherwise we duplicate the information.
* launch: keep pointer to current element in ParseContext
  This is in preparation for a refactoring of the error handling code. This
  way, we don't have to explicitly pass line number information around - we
  can instead pull it from the ParseContext when the error is generated.
* Merge pull request `#27 <https://github.com/xqms/rosmon/issues/27>`_ from xqms/feature/spec_tests
  roslaunch/XML spec unit tests
* launch: launch_config: error on <include clear_params="true" />
  Even the roslaunch/XML spec says this is "extremely dangerous". We will
  explicitly *not* support that one for now.
* launch: substitution_python: fix type deduction
  py::extract actually includes automatic conversion, so it is not
  appropriate for checking the returned object type. Use Python API instead.
* launch: handle <node clear_params="true"> attribute
* launch: launch_config: add support for <node cwd="..." />
* launch: launch_config: node uniqueness check should consider namespaces
* launch: substitution_python: report python exceptions more completely
* launch: launch_config: error if node name is not unique
* launch: launch_config: accept True/False as boolean values as well
  We are lenient here and accept the pythonic forms "True" and "False"
  as well, since roslaunch seems to do the same. Even the roslaunch/XML
  spec mentions True/False in the examples, even though they are not
  valid options for if/unless and other boolean attributes...
  http://wiki.ros.org/roslaunch/XML/rosparam
* launch_config: add check for invalid <param> combinations
* xml: param: test robustness against malformed tags
* launch: launch_config: propagate exceptions from lazy param threads
  .. to main thread.
* launch: launch_config: check if <param> commands exit normally
* launch_config: handle binfile
* launch: handle type "yaml" parameters (new in roslaunch since Lunar)
  This is actually a bit complicated, since this breaks a previous assumption
  we made: Our lazy evaluation of parameters depend on a 1:1 mapping of
  parameter names to jobs - this is not the case with YAML parameters, since
  one YAML file can turn into multiple params on the parameter server.
  So we handle YAML parameters separately from "ordinary" parameters, i.e.
  here our lazy evaluation does not prevent multiple loadings of the same
  parameters.
* cmake: basic rostest depends on rosmon target
  This makes sure that "make run_tests" also (re-)builds rosmon.
* launch: larger refactoring of param parsing
  Simplifies the forced type logic and applies it to "command" and "textfile"
  results as well.
* launch: split off as shared library and offer string parsing interface
  Preparation for more specific unit tests on roslaunch XML loading.

src/terminal.cpp
// Encapsulates terminal control (colors, cursor, ...)
// Author: Max Schwarz <max.schwarz@uni-bonn.de>

#include "terminal.h"

#include <termios.h>

#include <term.h>
#include <curses.h>

#include <cstdio>


#include <sys/ioctl.h>
#include <sys/types.h>

#include <boost/lexical_cast.hpp>
#include <boost/regex.hpp>
#include <boost/tokenizer.hpp>

namespace rosmon
{

Terminal::Parser::Parser()
 : m_state(STATE_ESCAPE)
 , m_fgColor(-1)
 , m_bgColor(-1)
 , m_bold(false)
{
}

void Terminal::Parser::parseSetAttributes(const std::string& str)
{
	using Tokenizer = boost::tokenizer<boost::char_separator<char>>;

	boost::char_separator<char> sep(";");
	Tokenizer tok(str, sep);

	for(Tokenizer::iterator it = tok.begin(); it != tok.end(); ++it)
	{
		errno = 0;
		auto endptr = const_cast<char*>(it->c_str());
		int code = strtoul(it->c_str(), &endptr, 10);

		if(errno != 0 || *endptr != 0)
		{
			// Error in specification, break out of here
			m_fgColor = -1;
			m_bgColor = -1;
			return;
		}

		if(code == 0)
		{
			m_fgColor = -1;
			m_bgColor = -1;
		}
		else if(code >= 30 && code <= 37)
			m_fgColor = code - 30;
		else if(code >= 40 && code <= 47)
			m_bgColor = code - 40;
		else if(code == 1)
			m_bold = true;
	}
}

void Terminal::Parser::parse(char c)
{
	switch(m_state)
	{
		case STATE_ESCAPE:
			if(c == '\033')
				m_state = STATE_TYPE;
			break;
		case STATE_TYPE:
			if(c == '[')
			{
				m_state = STATE_CSI;
				m_buf.clear();
			}
			else
				m_state = STATE_ESCAPE;
			break;
		case STATE_CSI:
			if(c == 'm')
			{
				parseSetAttributes(m_buf);
				m_state = STATE_ESCAPE;
			}
			else
			{
				m_buf.push_back(c);
				if(m_buf.length() >= 16)
					m_state = STATE_ESCAPE;
			}
			break;
	}
}

void Terminal::Parser::parse(const std::string& str)
{
	for(char c : str)
		parse(c);
}

void Terminal::Parser::apply(Terminal* term)
{
	if(m_fgColor >= 0 && m_bgColor >= 0)
		term->setSimplePair((SimpleColor)m_fgColor, (SimpleColor)m_bgColor);
	else
	{
		term->setStandardColors();
		if(m_fgColor >= 0)
			term->setSimpleForeground((SimpleColor)m_fgColor);
		else if(m_bgColor >= 0)
			term->setSimpleBackground((SimpleColor)m_fgColor);
	}
}

Terminal::Terminal()
 : m_valid(false)
 , m_256colors(false)
 , m_truecolor(false)
{
	// Override using environment variable
	char* overrideMode = getenv("ROSMON_COLOR_MODE");
	const char* termOverride = nullptr;
	if(overrideMode)
	{
		if(strcasecmp(overrideMode, "truecolor") == 0)
		{
			termOverride = "xterm-256color";
			m_256colors = true;
			m_truecolor = true;
		}
		else if(strcasecmp(overrideMode, "256colors") == 0)
		{
			termOverride = "xterm-256color";
			m_256colors = true;
			m_truecolor = false;
		}
		else if(strcasecmp(overrideMode, "ansi") == 0)
		{
			m_256colors = false;
			m_truecolor = false;
		}
		else
		{
			fprintf(stderr, "Warning: Unknown ROSMON_COLOR_MODE value: '%s'\n", overrideMode);
		}
	}
	else
	{
		// Detect truecolor-capable terminals
		if(getenv("KONSOLE_DBUS_SESSION"))
		{
			// Sadly, there is no way to determine the Konsole version. Since
			// any reasonably recent version supports true colors, just assume
			// true color support
			m_truecolor = true;
			m_256colors = true;
		}

		char* vte_version = getenv("VTE_VERSION");
		if(vte_version && boost::lexical_cast<unsigned int>(vte_version) >= 3600)
		{
			m_256colors = true;
			m_truecolor = true;
		}
	}

	int ret;
	if(setupterm(termOverride, STDOUT_FILENO, &ret) != OK)
	{
		printf("Could not setup the terminal. Disabling all colors...\n");
		return;
	}

	m_valid = true;

	if(!m_256colors && !overrideMode)
	{
		// Detect 256 color terminals
		int num_colors = tigetnum("colors");
		m_256colors = num_colors >= 256;
	}

	{
		const char* bgColor = tigetstr("setab");
		if(bgColor && bgColor != (char*)-1)
			m_bgColorStr = bgColor;
		else
			printf("Your terminal does not support ANSI background!\n");
	}
	{
		const char* fgColor = tigetstr("setaf");
		if(fgColor && fgColor != (char*)-1)
			m_fgColorStr = fgColor;
		else
			printf("Your terminal does not support ANSI foreground!\n");
	}

	m_opStr = tigetstr("op");
	m_sgr0Str = tigetstr("sgr0");
	m_elStr = tigetstr("el");
	m_upStr = tigetstr("cuu");

	m_boldStr = tigetstr("bold");
}

bool Terminal::has256Colors() const
{
	return m_256colors;
}

void Terminal::setCursorInvisible()
{
	if(!m_valid)
		return;

	putp(tigetstr("civis"));
}

void Terminal::setCursorVisible()
{
	if(!m_valid)
		return;

	putp(tigetstr("cnorm"));
}

static int ansiColor(uint32_t rgb)
{
	int r = (rgb & 0xFF);
	int g = (rgb >> 8) & 0xFF;
	int b = (rgb >> 16) & 0xFF;

	r = r * 6 / 256;
	g = g * 6 / 256;
	b = b * 6 / 256;

	return 16 + 36 * r + 6 * g + b;
}

void Terminal::setBackgroundColor(uint32_t color)
{
	if(!m_valid)
		return;

	if(m_truecolor)
	{
		char buf[256];
		snprintf(buf, sizeof(buf), "\033[48;2;%d;%d;%dm", color & 0xFF, (color >> 8) & 0xFF, (color >> 16) & 0xFF);
		fputs(buf, stdout);
	}
	else
	{
		char* out = tiparm(m_bgColorStr.c_str(), ansiColor(color));
		putp(out);
	}
}

void Terminal::setForegroundColor(uint32_t color)
{
	if(!m_valid)
		return;

	if(m_truecolor)
	{
		char buf[256];
		snprintf(buf, sizeof(buf), "\033[38;2;%d;%d;%dm", color & 0xFF, (color >> 8) & 0xFF, (color >> 16) & 0xFF);
		fputs(buf, stdout);
	}
	else
	{
		char* out = tiparm(m_fgColorStr.c_str(), ansiColor(color));
		putp(out);
	}
}

void Terminal::setEcho(bool on)
{
	// Switch character echo on
	termios ios;
	if(tcgetattr(STDIN_FILENO, &ios) == 0)
	{
		if(on)
		{
			ios.c_lflag |= ECHO;
			ios.c_lflag |= ICANON;
		}
		else
		{
			ios.c_lflag &= ~ECHO;
			ios.c_lflag &= ~ICANON;
		}

		tcsetattr(STDIN_FILENO, TCSANOW, &ios);
	}
}

void Terminal::setBold(bool on)
{
	if(!m_valid)
		return;

	if(on)
		putp(m_boldStr.c_str());
}

void Terminal::setSimpleForeground(SimpleColor color)
{
	if(!m_valid)
		return;

	char* out = tiparm(m_fgColorStr.c_str(), color);
	putp(out);
}

void Terminal::setSimpleBackground(SimpleColor color)
{
	if(!m_valid)
		return;

	char* out = tiparm(m_bgColorStr.c_str(), color);
	putp(out);
}

void Terminal::setSimplePair(SimpleColor fg, SimpleColor bg)
{
	if(!m_valid)
		return;

	setSimpleForeground(fg);
	setSimpleBackground(bg);
}

void Terminal::setStandardColors()
{
	if(!m_valid)
		return;

	putp(m_opStr.c_str());
	putp(m_sgr0Str.c_str());
}

void Terminal::clearToEndOfLine()
{
	if(!m_valid)
		return;

	putp(m_elStr.c_str());
}

void Terminal::moveCursorUp(int numLines)
{
	if(!m_valid)
		return;

	putp(tparm(m_upStr.c_str(), numLines));
}

void Terminal::moveCursorToStartOfLine()
{
	putchar('\r');
}

bool Terminal::getSize(int* outColumns, int* outRows)
{
	struct winsize w;
	if(ioctl(STDIN_FILENO, TIOCGWINSZ, &w) == 0)
	{
		*outColumns = w.ws_col;
		*outRows = w.ws_row;
		return true;
	}
	else
		return false;
}

void Terminal::setWindowTitle(const std::string& title)
{
	char buf[256];

	// Konsole style
	snprintf(buf, sizeof(buf), "\033]30;%s\007", title.c_str());
	fputs(buf, stdout);

	// screen/tmux style
	snprintf(buf, sizeof(buf), "\033k%s\033\\", title.c_str());
	fputs(buf, stdout);
}

void Terminal::clearWindowTitle(const std::string& backup)
{
	char buf[256];

	fputs("\033]30;%d : %n\007", stdout);

	// screen/tmux style
	snprintf(buf, sizeof(buf), "\033k%s\033\\", backup.c_str());
	fputs(buf, stdout);
}

}
